### PR TITLE
Multiple choice with 1 or 2 answers

### PR DESCRIPTION
**Pull request by breville**
Multiple choice now supports puzzles with 1 or 2 "right" answers.  In the latter case, both right answers must be selected simultaneously.

When a .multi file has a type (which is optional) of 'multi2' it can have two right answers.

The UI has been updated to use font-awesome glyphs instead of radio buttons.  

When doing a 'multi2' puzzle, when two items are already selected, clicking on a new item will deselect the last-recently clicked of the other two items.  It's also possible to click a selected item to deselect it.

A new warning dialog pops when not enough items have been selected in 'multi2' puzzles.

### CODE REVIEW

File: dashboard/app/views/levels/_multi.html.haml [hunk 1]
```diff
@@ -53,75 +55,175 @@
...
-    // Pre-select previously submitted response if available
-    if (selectionIndex != -1) {
-      var radioButton = $('#choice_' + selectionIndex);
-      radioButton.prop('checked', true);
-      var correctChoice = $(radioButton).attr('correct') === 'true';
-      selection = correctChoice ? "right" : "wrong";
-      enableButton(true);
+    // Pre-select previously submitted response if available.
+    if (previousResult.length >= 1)
+    {
```
Comment (joshlory) [inline dashboard/app/views/levels/_multi.html.haml]: Nit: `{` on previous line.

File: dashboard/app/views/levels/_multi.html.haml [hunk 2]
```diff
@@ -53,75 +55,175 @@
...
     }
   });
+
+  function validateAnswers()
+  {
+    if (selectedAnswers.length == numAnswers)
+    {
+      for (var i = 0; i < numAnswers; i++)
+      {
+        if (! answers[selectedAnswers[i]])
```
Comment (joshlory) [inline dashboard/app/views/levels/_multi.html.haml]: Nit: no space after `!`.

**Comment by breville** [comment]
thanks @joshlory .. you'll notice the whole file has a different code format because it was written a while ago.. i'm going to avoid layout changes in this PR just so it can focus on code changes.  hope that's okay.

**Comment by joshlory** [comment]
Ah that makes sense, no worries.

**Comment by wjordan** [comment]
I looked through the code, and nothing obviously problematic jumps out at me. I think it looks good for a minimal, incremental new feature on an existing chunk of code we haven't touched much. Two points I think are worth noting but do not necessarily block this PR, since spending time on them might not be too practical for this incremental change:

- This feature lacks automated tests (unit or UI), and the tight coupling between logic and presentation in the haml templates makes it rather difficult to try and follow the expected behavior by just looking at the code, so I wasn't able to confidently confirm that this PR actually works. I trust you've manually tested the new behavior yourself and that it functions as expected. Maybe worth checking `multi.feature` at least to make sure this change doesn't break the existing test, and also to see if the new multi-answer functionality could be tested there?

- The addition of a `type` parameter to distinguish between single-answer multi and multi-answer multi is slightly at odds with the existing class-based inheritance model in `Level` (which is also called `type`). I assume the reason why inheritance isn't used here (e.g., `class MultiMulti < Multi`) is because of the extra few boilerplate steps needed to set up a new Level type, compared to the simpler process of adding/changing a single property on an existing Level type. I think it's manageable with a small, incremental change like this, but I just wanted to point out the potential for confusion/duplication there so we don't let it become a burden down the road.

**Comment by breville** [comment]
Thanks Will, this is great feedback.  

I have tested a bunch manually but have added to my to-do list to add an automated UI test.  Is there a better name for 'type' here?

Also, I felt the same way about the coupling of UI & code.. this would be a great learner project to rewrite in React with some spare bandwidth.

**Pull request merged**